internal/ratelimit/handler.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package ratelimit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"github.com/hashicorp/boundary/internal/types/action"
+	"github.com/hashicorp/boundary/internal/types/resource"
+	rate "github.com/hashicorp/go-rate"
+)
+
+var pathRegex = regexp.MustCompile(`/v1/(?P<resource>[\w-]+)((/(?P<id>[\w/]+))?(:(?P<action>[\w-:]+)?)?)?`)
+
+func extractResourceAction(path, method string) (res, act string, err error) {
```

Comment:
I sort of like @mgaffney 's suggestion to return defined error codes which can
be used by the caller (the handler) to return the appropriate error/status.

Reply:
See: https://github.com/hashicorp/boundary/pull/3825#discussion_r1376685709